### PR TITLE
[BEAM-6077] If available, use max_parallelism for splitting unbounded source

### PR DESCRIPTION
**Pull request by JozoVilcek**
Allow to set max parallelism to FlinkRunner and leverage it for UnboundedSourceWrapper to produce finer grained state which will allow to re-scale job up to max-parallelism. 
If max parallelism is not set, system behaves like before this PR and splits according to parallelism.

@mxm pointed out that there is `BoundedToUnboundedSourceAdapter` which can lead to unnecessary splitting e.g. for files. Therefore, in `FlinkStreamingTransformTranslators`, max parallelism is used only for `UnboundedReadSourceTranslator` and not for `BoundedReadSourceTranslator` which uses this adapter. 

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by JozoVilcek** [comment]
By `simple unit test which tests the setting of parallelism on the source` you mean somewhere around `ReadSourceTranslator` that if max parallelism is set that is pushed to the source wrapper and if not then it is "normal" parallelism? Or do you mean something else?

**Comment by mxm** [comment]
Yes, that's what I meant.

**Comment by JozoVilcek** [comment]
I did manage to put together some tests. Let me know if you would like to chanage them somehow to be e.g. less integration like.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java [hunk 1]
```diff
@@ -1284,6 +1284,10 @@ public String getUrn(CreateStreamingFlinkView.CreateFlinkPCollectionView<?, ?> t
 
     private final UnboundedSourceWrapper<OutputT, CheckpointMarkT> unboundedSourceWrapper;
 
+    UnboundedSourceWrapper<OutputT, CheckpointMarkT> getUnderlyingSource() {
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java]: Would add `@VisibleForTesting`.

File: runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslatorsTest.java [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+  @Test
+  public void readSourceTranslatorBoundedWithMaxParallelism() {
+
+    final int maxParallelism = 6;
+    final int parallelism = 2;
+
+    Read.Bounded transform = Read.from(new TestBoundedSource(maxParallelism));
+    StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+    env.setParallelism(parallelism);
+    env.setMaxParallelism(maxParallelism);
```
Comment (mxm) [inline runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslatorsTest.java]: Could we also add another test which verifies the max parallelism is applied during translation from FlinkPipelineOptions to the ExecutionEnviroment?

**Comment by mxm** [comment]
Run Portable_Python PreCommit

**Pull request merged**